### ISSUE TITLE
No such file or directory at the End of Workflow using NFS

### ISSUE THREAD

**Issue by Guigzai** [DataBiosphere/toil#4092]
Hi,

We are using a cluster manage by Mesos.
Toil Leader run on virtual machine that mounting /shared with NFS.
Toil Executor run on computing node that mounting /shared with GPFS.

It often occur this problem at the end of workflow :
```
[2022-05-01T18:10:43+0000] [MainThread] [I] [toil] Running Toil version 5.6.0-c34146a6437e4407a61e946e968bcce67a0ebbca on host 7ddc344a6a16.
[2022-05-01T18:10:43+0000] [MainThread] [D] [toil] Configuration: {'workflowID': 'c7be9427-34e4-46d4-bdc5-06985b26f890', 'workflowAttemptNumber': 0, 'jobStore': 'file:/shared/job_stores/workflow-jobstore', 'logLevel': 'DEBUG', 'workDir': None, 'noStdOutErr': False, 'stats': False, 'clean': 'never', 'clusterStats': None, 'restart': False, 'batchSystem': 'mesos', 'disableAutoDeployment': True, 'maxLocalJobs': 8, 'manualMemArgs': False, 'coalesceStatusCalls': False, 'statePollingWait': None, 'parasolCommand': 'parasol', 'parasolMaxBatches': 10000, 'scale': 1.0, 'mesos_endpoint': 'chronos-scheduler:5050', 'allocate_mem': False, 'tes_endpoint': 'http://xxx.xxx.xxx.xxx:8000', 'tes_user': None, 'tes_password': None, 'tes_bearer_token': None, 'runCwlInternalJobsOnWorkers': False, 'disableCaching': True, 'linkImports': True, 'moveExports': True, 'provisioner': None, 'nodeTypes': [], 'minNodes': None, 'maxNodes': [10], 'targetTime': 1800, 'betaInertia': 0.1, 'scaleInterval': 60, 'preemptableCompensation': 0.0, 'nodeStorage': 50, 'nodeStorageOverrides': [], 'metrics': False, 'maxPreemptableServiceJobs': 9223372036854775807, 'maxServiceJobs': 9223372036854775807, 'deadlockWait': 60, 'deadlockCheckInterval': 30, 'defaultMemory': 2147483648, 'defaultCores': 1, 'defaultDisk': 2147483648, 'readGlobalFileMutableByDefault': False, 'defaultPreemptable': False, 'maxCores': 9223372036854775807, 'maxMemory': 9223372036854775807, 'maxDisk': 9223372036854775807, 'retryCount': 1, 'enableUnlimitedPreemptableRetries': False, 'doubleMem': False, 'maxJobDuration': 9223372036854775807, 'rescueJobsFrequency': 60, 'environment': {}, 'disableChaining': False, 'disableJobStoreChecksumVerification': False, 'maxLogFileSize': 0, 'writeLogs': None, 'writeLogsGzip': None, 'writeLogsFromAllJobs': False, 'sseKey': None, 'servicePollingInterval': 60, 'useAsync': True, 'forceDockerAppliance': False, 'statusWait': 3600, 'disableProgress': False, 'debugWorker': False, 'disableWorkerOutputCapture': False, 'badWorker': 0.0, 'badWorkerFailInterval': 0.01, 'cwl': False, 'cleanWorkDir': 'always'}
[2022-05-01T18:10:43+0000] [MainThread] [D] [toil.realtimeLogger] Real-time logging disabled


[2022-05-01T18:15:17+0000] [MainThread] [D] [toil.batchSystems.mesos.batchSystem] Stopping Mesos driver
[2022-05-01T18:15:17+0000] [MainThread] [D] [toil.batchSystems.mesos.batchSystem] Joining Mesos driver
[2022-05-01T18:15:17+0000] [MainThread] [D] [toil.batchSystems.mesos.batchSystem] Joined Mesos driver
[2022-05-01T18:15:17+0000] [MainThread] [D] [toil.common] ... finished shutting down the batch system in 0.017150402069091797 seconds.
Traceback (most recent call last):
  File "/usr/local/bin/toil-cwl-runner", line 8, in <module>F
    sys.exit(main())
  File "/usr/local/lib/python3.9/site-packages/toil/cwl/cwltoil.py", line 3486, in main
    outobj = toil.start(wf1)
  File "/usr/local/lib/python3.9/site-packages/toil/common.py", line 951, in start
    return self._runMainLoop(rootJobDescription)
  File "/usr/local/lib/python3.9/site-packages/toil/common.py", line 1273, in _runMainLoop
    return Leader(config=self.config,
  File "/usr/local/lib/python3.9/site-packages/toil/leader.py", line 251, in run
    self.innerLoop()
  File "/usr/local/lib/python3.9/site-packages/toil/leader.py", line 720, in innerLoop
    self._gatherUpdatedJobs(updatedJobTuple)
  File "/usr/local/lib/python3.9/site-packages/toil/leader.py", line 678, in _gatherUpdatedJobs
    self.processFinishedJob(bsID, exitStatus, wallTime=wallTime, exitReason=exitReason)
  File "/usr/local/lib/python3.9/site-packages/toil/leader.py", line 1107, in processFinishedJob
    self.toilState.reset_job(jobStoreID)
  File "/usr/local/lib/python3.9/site-packages/toil/toilState.py", line 148, in reset_job
    new_truth = self.__job_store.load_job(job_id)
  File "/usr/local/lib/python3.9/site-packages/toil/jobStores/fileJobStore.py", line 213, in load_job
    with open(jobFile, 'rb') as fileHandle:
FileNotFoundError: [Errno 2] No such file or directory: '/shared/job_stores/workflow-jobstore/jobs/kind-CWLJob/instance-t1swvksq/job'
```

Thanks for your help
Best Regards
Guigzaï

┆Issue is synchronized with this [Jira Story](https://ucsc-cgl.atlassian.net/browse/TOIL-1163)
┆friendlyId: TOIL-1163

**Comment by adamnovak** [comment]
It looks like the problem is happening here:

https://github.com/DataBiosphere/toil/blob/a98acdb5cbe0f850b2c11403d147577d9971f4e1/src/toil/leader.py#L1115-L1119

At the first line we call `job_exists`, and at that point in time the file `/shared/job_stores/workflow-jobstore/jobs/kind-CWLJob/instance-t1swvksq/job` is still visible to the leader, because the completion message from the worker to the leader via Mesos has arrived before the worker's deletion of the file is visible on the filesystem.

But by the time we get to the `reset_job` call, that file's deletion has propagated to the leader node, and we can't open the file anymore. We are supposed to actually handle this case already here:

https://github.com/DataBiosphere/toil/blob/a98acdb5cbe0f850b2c11403d147577d9971f4e1/src/toil/toilState.py#L147-L149

But we are getting a `FileNotFoundError` and not the `NoSuchJobException` we are expecting.

That's because the `FileJobStore`'s `load_job` implementation isn't actually safe against the race condition that is happening here. It needs to throw `NoSuchJobException` and never `FileNotFoundError` when a job is missing, but it accomplishes this by first checking for the job, throwing `FileNotFoundError` if it doesn't exist, and then opening it. If, as in this case, the deletion of the job becomes visible between the check and the open, the open can fail and throw a `FileNotFoundError`.

https://github.com/DataBiosphere/toil/blob/a98acdb5cbe0f850b2c11403d147577d9971f4e1/src/toil/jobStores/fileJobStore.py#L209-L214

**So, one change we should make here is to catch the `FileNotFoundError` instead of polling before the open, in `load_job`. That would prevent this particular crash.**

But, the underlying problem is that the filesystem is willing to claim that the file exists after it is possible for a message to have arrived via Mesos from another machine that has deleted the file. Toil really wants a filesystem that can make consistency guarantees strong enough to prevent this from happening: it needs to appear consistent even when nodes can communicate in ways other than the filesystem itself. A deletion operation should not finish until after the point in real time when other readers are no longer able to see the file, even if they have e.g. cached the file's existence. Your filesystem does not make this guarantee; it has probably been turned off for speed, or in the case of GPFS might not be available in any configuration. We do try to work around the resulting problems, but we'll never be able to guarantee correct operation of Toil against this filesystem setup, if the filesystem can keep us waiting arbitrarily long to see the changes that were made by jobs that finished.

The workaround here is to wait for the filesystem to settle and `--restart` the workflow; it should be able to continue where it left off. The other workaround is to reconfigure your NFS/GPFS for strong(er) consistency.

**Comment by Guigzai** [comment]
Do you think about use message broker in alternative of file system to manage workflow ?
A other idea is to separate JobStore in to part : ControlJobstore and DataJobstore by example.
In this way we could : 
 * Use ControlJobstore with fast IO or in memory (it is to have a hight performance space to read/write control files that manage workflows and like that reduce the risk of bad sync),
 * Use DataJobstore to store data file during the workflow execution.

Regards

**Comment by adamnovak** [comment]
I don't think we've looked at using a message broker for anything. It's not clear that a message concept quite fits what an outstanding Toil job is.

We did something like that two-part model with the AWS job store: keep large objects in S3, and small objects (mostly jobs) in SimpleDB.

But we're moving that to use only S3 and just keep everything there, since S3 has the consistency guarantees we need now, and SimpleDB is out of style.

I'm also not sure that that model quite solves the problem: the DataJobstore still really needs to be consistent, if we don't want to be spin-waiting for files to exist or risking using old versions of updated files. Though it might be easier for it to win the races it is in if we don't hit it so quickly.
